docs_src/design/adr/application/0025-Record-and-Replay.md

```diff
@@ -0,0 +1,128 @@
+
+The replay async Go function will use the [Background Publishing](https://docs.edgexfoundry.org/2.3/microservices/application/AdvancedTopics/#background-publishing) capability to send the recorded Events to the EdgeX MessageBus using the same publish topic scheme used by Device Services, which is `edgex/events/device/<device-profile-name>/<device-name>/<source-name>`. The App SDK has the  [Publish Topic Placeholders](https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#publish-topic-placeholders) capability built-in to facilitate this. The data for these topics is available from the Event DTO. The timestamps in the Events and Readings published will be set to the current date/time. This requires a copy be made of the Event/Readings as they are published in order to not corrupt the original data.
+
+Once the first event is published the replay function will calculate the wait time to use before sending the next Event from the recorded data. This will be based on the time difference from the original timestamp of the previous event published and the timestamp of the next event multiplied by the inverse of the `Replay Rate` specified in the request DTO. 
+
+!!! example - "Examples - Replay Rate wait time calculation"
+    Delta time between original Events is 800ms<br/>
+    Replay rate is 2.0 (100% faster) making wait time 400ms (800ms * (1 / 2.0))<br/>
+    Replay rate is 0.5 (100% slower) making wait time 1600ms (800ms * (1 / 0.5))
+
+The replay function will repeat publishing the recorded data per the `Repeat Count` in from the DTO. 
+
+#### Replay Request DTO
+
+##### Replay Rate
+
+Required rate at which to replay the data compared to the rate the data was recorded. Float value greater than 0 where 1 is the same rate, less than 1 is slower rate and greater than 1 is faster rate than the rate the data was recorded. 
+
+##### Repeat Count
+
+Optional count of number of times to repeat the replay. Default is 1 if not specified or value is 0.
+
+### Download endpoint (Export)
+
+This `GET` API will request that the previously recorded data be exported as a file download. It will accept an optional query parameter to specify compression (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.
+
+The file content will be the [Recorded Data DTO](#recorded-data-dto) as define below. The request handler will build the DTO described below by extracting the recorded `Events` from in-memory storage, pulling the referenced `Device Profiles` and `Devices` from Core Metadata using the names from in-memory storage. The file extension used will be `.json`, `.zip` or `.gzip` depending on the compression selected.
```

Comment:
More of an implementation concern: Data collection in memory, compression, and serialization together will be a recipe for memory exhaustion. I suggest making use of the secondary storage and data streaming techniques for some if not all the operations. The data can be streaming to an HTTP client too; see chunked transfer encoding: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding.

Reply:
> I suggest making use of the secondary storage
Secondary storage can be an enhancement (walk phase) once we get some usage and feedback on the need.

Reply:
> I suggest making use of the secondary storage and data streaming techniques for some if not all the operations. The data can be streaming to an HTTP client too; see chunked transfer encoding: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding.

These are implementation details which we'll consider during the implementation.